hermes/probe/journal/journal.go

```diff
@@ -0,0 +1,364 @@
+	if journal.Intent == nil {
+		return nil
+	}
+
+	fileID, err := strconv.Atoi(journal.Intent.Filename[len("Hermes_"):len("Hermes_ID")])
+	if err != nil {
+		return probe.NewProbeError(metrics.ProbeFailed, err)
+	}
+
+	switch journal.Intent.FileOperation {
+	case pb.Intent_CREATE:
+		// TODO(evanSpendlove): Check if the file with this filename has already been created. Check the filesystem for the file existing and check the filename is valid.
+		if journal.Filenames[int32(fileID)] != journal.Intent.Filename {
+			// TODO(evanSpendlove): Call createFile() passing the fileID and return
+			logger.Debug("File create not complete yet")
+			journal.Filenames[int32(fileID)] = journal.Intent.Filename
+		}
+	case pb.Intent_DELETE:
+		if name := journal.Filenames[int32(fileID)]; name != "" {
+			// TODO(evanSpendlove): Call  DeleteFile() passing the fileID and return
+			logger.Debug("File delete not complete yet")
+			delete(journal.Filenames, int32(fileID))
+		}
+	}
+	journal.Intent = nil
+	return nil
+}
+
+// UpdateJournal writes the current in-memory journal proto a new nil file on
+// the target system. It then deletes the old nil file located on the system.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+// TODO(evanSpendlove): Add existing nil file deletion to here
+func UpdateJournal(ctx context.Context, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	contents, err := proto.Marshal(target.Journal)
+	if err != nil {
+		return probe.NewProbeError(metrics.ProbeFailed, err)
+	}
+
+	file := &journalFile{
+		contents: contents,
+	}
+
+	if err := file.checksumContents(); err != nil {
+		return err
+	}
+
+	file.name = fmt.Sprintf("%s%x", journalFilePrefix, file.checksum)
+	bucket := target.Target.GetBucketName()
+	logger.Debug("UpdateNilFile(): in-memory journal successfully marshalled, proceeding to upload to target storage system.")
+
+	var writer stiface.Writer
+	if err := RecordAPILatency(target.LatencyMetrics.APICallLatency, metrics.APICreateFile, func() *probe.ProbeError {
+		writer = client.Bucket(bucket).Object(file.name).NewWriter(ctx)
+		length, err := io.Copy(writer, file.reader())
+		if err != nil {
+			switch err {
+			case storage.ErrBucketNotExist:
+				return probe.NewProbeError(metrics.BucketMissing, err)
+			default:
+				return probe.NewProbeError(metrics.ProbeFailed, err)
+			}
+		}
+		if length != int64(len(file.contents)) {
+			return probe.NewProbeError(metrics.APICallFailed, fmt.Errorf("UpdateNilFile(): failed to write correct number of bytes to file, expected %d, got %d", len(file.contents), length))
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if err := writer.Close(); err != nil {
```

Comment:
Shouldn't the writer be closed in any case?
What if there's an error and we return on line 338?

Reply:
I've moved this to inside the closure so that an attempt to close the writer, which might fail, is always made.